relay-general/src/protocol/breakdowns.rs

```diff
@@ -0,0 +1,339 @@
+use std::collections::HashMap;
+use std::ops::{Deref, DerefMut};
+
+use serde::{Deserialize, Serialize, Serializer};
+
+use crate::protocol::{Event, Measurement, Measurements, Timestamp};
+use crate::types::{Annotated, Error, FromValue, Object, SkipSerialization, ToValue, Value};
+
+#[derive(Clone, Debug)]
+struct TimeWindowSpan {
+    start_timestamp: Timestamp,
+    end_timestamp: Timestamp,
+}
+
+impl TimeWindowSpan {
+    fn new(start_timestamp: Timestamp, end_timestamp: Timestamp) -> Self {
+        if end_timestamp < start_timestamp {
+            return TimeWindowSpan {
+                start_timestamp: end_timestamp,
+                end_timestamp: start_timestamp,
+            };
+        }
+
+        TimeWindowSpan {
+            start_timestamp,
+            end_timestamp,
+        }
+    }
+
+    fn get_duration(&self) -> u64 {
+        (self.end_timestamp.timestamp_nanos() - self.start_timestamp.timestamp_nanos()).abs() as u64
+    }
+}
+
+#[derive(PartialEq, Eq, Hash)]
+enum OperationBreakdown<'op_name> {
+    Emit(&'op_name str),
+    DoNotEmit(&'op_name str),
+}
+
+type OperationNameIntervals<'op_name> = HashMap<OperationBreakdown<'op_name>, Vec<TimeWindowSpan>>;
+
+fn get_op_time_spent(mut intervals: Vec<TimeWindowSpan>) -> Option<u64> {
+    if intervals.is_empty() {
+        return None;
+    }
+
+    // sort by start_timestamp in ascending order
+    intervals.sort_unstable_by(|a, b| a.start_timestamp.partial_cmp(&b.start_timestamp).unwrap());
```

Comment:
_nit_: You can also use `sort_unstable_by_key` for brevity and to get rid of the `unwrap()`:

```suggestion
    intervals.sort_unstable_by_key(|span| span.start_timestamp);
```